Source/Public/Set-VSTeamDefaultProject.ps1

```diff
@@ -49,19 +49,19 @@ function Set-VSTeamDefaultProject {
 
    process {
       if ($Force -or $pscmdlet.ShouldProcess($Project, "Set-VSTeamDefaultProject")) {
-         if (_isOnWindows) {
-            if (-not $Level) {
-               $Level = "Process"
-            }
+         #Set at the process level, for any OS
+         $env:TEAM_PROJECT = $Project
+         [vsteam_lib.Versions]::DefaultProject = $Project
 
-            # You always have to set at the process level or they will Not
-            # be seen in your current session.
-            $env:TEAM_PROJECT = $Project
-            [vsteam_lib.Versions]::DefaultProject = $Project
+         $env:TEAM_PROCESS = [vsteam_lib.Versions]::DefaultProcess = Get-VSTeamProcess  -ExpandProjects |
+               Where-Object Projects -Contains $Project | Select-Object -ExpandProperty Name
```

Comment:
As I mentioned in the comment of the PR I am not a fan of $env:TEAM_PROCESS. Even if I was this is not where it should be getting set. Why would anyone call Set-VSTeamDefaultProject and expect the Process to be set?

Reply:
OK. We're going to have work though a difference in style and ideas of what's best practice. 
`PSDefaultParameterValues["*-vsteam*:Something"]`   Can be problematic 

If the code has 
`Param  ( [type]$something = [class]::member) `
or 
`Param  ( [type]$something = $Scope:Variable ) `
where scope is` Env: `, `Script:,` or` Global:` 
Then that variable will be set.  And we and can distinguished between _user-passed_ and _set-to-default_ by looking at `$PSBoundParameter.containskey('Something')`   _defaulted_, parameters are not _bound_ (btw the common trick of treating `$something` as a boolean to see if was passed fails if the value converts to `false` (for example it's an integer or an Enum, and 0 is a valid value so I commonly check this way).

Now I have two ways I want to ask for WorkItem Types.  Give me the types for the a project, (the current project by default) or give me the types for a specific process template. (Scrum, agile etc)

This commandline should be impossible. 
`Get-VsteamWorkItemType  -ProjectName  MyScrumProject  -ProcessTemplate MyAgileTemplate`

So one would write 
```
Param (
   [Parameter(ParameterSetName='ByProject')]
   $ProjectName  
   [Parameter(ParameterSetName='ByTemplate')]
   $ProcessTemplate
```

So now the user can't type both
`Get-VsteamWorkItemType -ProcessTemplate MyAgileTemplate`
Now causes an error, because `PSDefaultParameterValues["*-vsteam*:ProjectName  "]`  **always** stuffs the current project name in. So in this case I have had to allow a parameter set which takes BOTH and ignores `ProjectName `if there is a user supplied  `ProcessTemplate` 

**Ideally Parameter defaults should be set where the parameter is defined** (and the help should say what the default is), there's obviously a big lump of work to change this for `ProjectName` parameters but I did want to get it right for new things.  
There are other places where it needs background knowledge to understand defaults are being set. @SebastianSchuetze  told me that _callapi as a default for contentType 
it doesn't
```
      [string]$OutFile,
      [string]$ContentType,
      [string]$ProjectName,
```
Buried in the function there is this 
```
      $params = $PSBoundParameters
      if ($false -eq $PSBoundParameters.ContainsKey("ContentType")) {
         $params.Add('ContentType', 'application/json; charset=utf-8')
      }
``` 

To answer
> Why would anyone call Set-VSTeamDefaultProject and expect the Process to be set?

I don't think they would.  It is just **caching "Process of the current project".**  Otherwise when any command which works on process template is called either the user must supply the template name, or if no template is given the code needs to fetch all templates and find out which one the current project is in.  I **can** do that by `PSDefaultParameterValues["*-vsteam*:Something"]`   if you understand the issue and still prefer that way.

Reply:
Additional it did seem unnecessary to add the default process template to the user or system environment variables,; It is set in the local environment when connecting to a project and a situation where the user changes one but not the other isn't good.. 
I'm also wondering what happens if the user changes `PSDefaultParameterValues["*-vsteam*:Something"] ` without calling `set-VSTeamDefaultProject. `

Reply:
Sorry for the delay.  There is a lot to unpack here. I will try to take them one at a time. 

1. No matter if we decide to use environment variables or PSDefaultParametersValues there needs to be a Set-VSTeamDefaultProcessTemplate and Clear-VSTeamDefaultProcessTemplate. I don't want Set-VSTeamDefaultProject to have side effects.

2. I agree with you on the contentType default value on _callAPI and will address. 

3. Get-VsteamWorkItemType -ProjectName MyScrumProject -ProcessTemplate MyAgileTemplate should be impossible. I don't see how we set the default project affecting this. If we use the environment variable way or the PSDefaultParametersValues if a default is set for ProjectName you should get the error.

Reply:
I am ok with finding an alternative to using PSDefaultParametersValues. So your method is fine but I still want separate functions for setting and clearing that value to be consistent with DefaultProject.

Reply:
Don't worry about the delay. I worked at Microsoft a decade ago and know what it's like doing a big PowerShell thing as a side project (first PowerShell support for Hyper-V, when 2008 didn't have a module for it... that was me).   

1. I understand the not wanting side effects thing, but the idea is to have the default process template as _process for the current project_ and I'm just caching it instead of querying for it.  
Just checking my thinking the default process template _should_ follow the default project , shouldn't it ?  It would be quite easy don't change the process when the project changes under some conditions. 
 
In a way it would be better to have a `-DefaultProject` parameter for `Set-VSTeamAccount` so the job is done in one command instead of two/three and then have setting the project or process as individual commands which only changed one variable.   

2. :-)  

3. That's what most people would expect. But PSDefaultParameterValues  _fakes_ a command line parameter.  Here's a little sample
```
function test {
[cmdletbinding(DefaultParameterSetName="one")]
param (
    [parameter(mandatory=$true, ParameterSetName="one")]
    $param_one ,
    [parameter(ParameterSetName="two")]
    $param_two
)
$PSBoundParameters
$param_one
$param_two
}
```
on its own  `test`  prompts
`Supply values for the following parameters:`   
`param_one: `   

`test -param_two 2` or   `test -param_one 1` work as expected and `test -param_one 1 -param_two 2` throws a parameter error. If you then do 
`$PSDefaultParameterValues['Test:param_one']="Alpha" `     
`test`  works with no parameters on the command line and reports that `Param_one` was a bound parameter. 

What Default Parameter Values does is to fake  `-param_one "Alpha"` being on the commandline so if you type    
`test -param_two 2`  the command will always see `test -param_two 2 -param_one "Alpha"` which isn't valid. 

But if you write the function like this :
```
function test2 {
[cmdletbinding(DefaultParameterSetName="one")]
param (
    [parameter(ParameterSetName="one")]
    $param_one = $global:Par1, 
    [parameter(ParameterSetName="two")]
    $param_two
)
$PSBoundParameters
$param_one
$param_two
}
```
`Test2` will prompt for input unless you do  `$global:par1 = "Alpha"` first  and it works with a either parameter but not both. 
However when the value is defaulted in, it is _not a bound parameter_ so `test` on its own shows $param_one is "Alpha" but there are no bound parameters, and `test -param_two  2` shows $param_two has a value, `psBoundParameters` contains Param_two = 2 and the **variable**  $param_one has its default value even though there is no **parameter** param_one (it's the opposite of dynamic parameters, they go in PSBoundParameters but aren't automatically variables). 

I can think of 3 reasons for avoiding using `PSDefaultParameterValues` with wildcards - putting the parameter in when it isn't wanted is one, not having something in the function's parameter block telling you how a default gets there is a second (see #2 :-)  ) and another script might reset it.

Reply:
> In a way it would be better to have a `-DefaultProject` parameter for `Set-VSTeamAccount` so the job is done in one command instead of two/three and then have setting the project or process as individual commands which only changed one variable.

Adding a -DefaultProject parameter to Set-VSTeamAccount you would lose tab completion and validation because you have not authorized the account.

Reply:
Now we could add Set-VSTeamDefault and that way we only change that one function as we discover new defaults we want to provide. We need to leave Set-VSTeamDefaultProject for a while to allow a transfer from one to the other.

Reply:
> Adding a -DefaultProject parameter to Set-VSTeamAccount you would lose tab completion and validation because you have not authorized the account.

Yes, but I think that would be OK. Currently I have 
```Set-VSTeamAccount -Account $account  -SecurePersonalAccessToken (Import-Clixml .\myToken.xml) ;
Set-VSTeamDefaultProject $projectName```

In a file 
And I'm trying to avoid 
`Set-VSTeamDefaultProcess ((Get-VSTeamProcess -ExpandProjects) | where projects -Contains $projectname)`

In an ideal world I'd have a utility to create _credential_ where the user name was Account\project and the password was the access token (utility would get this from the clipboard so it's never in the history) and export it with `export-cliXML`   to vsteamCred.xml in my home directory  (a secure string saved like that can only be decoded by the person who saved it)  then  
`Connect-VSTeam  -credFile mycred.xml) ` 
uses import-clixml to get the cred,  splits name at the \ to get the account deals with the secure string and logs on as now; if there is anything after the \ it calls `Set-VSTeamDefaultProject` with that.  And either `Set-VSTeamDefaultProject` or a call to new code in either in a new `Set-VSTeamDefaultProcess` or inside this `connect` command would  set the process default.  As a default I'd have `home\vsteamCred.xml` as the path so can just run Connect-Vsteam (or some alias). And for those who don't want a credential file I'd allow credential input so they paste an access token but not into the command line.